CruCentralCoast/Community/CommunityGroups/CommunityGroupCell.swift

```diff
@@ -0,0 +1,70 @@
+//
+//  CommunityGroupCell.swift
+//  CruCentralCoast
+//
+//  Created by Landon Gerrits on 5/22/18.
+//  Copyright © 2018 Landon Gerrits. All rights reserved.
+//
+
+import UIKit
+
+class CommunityGroupCell: UITableViewCell {
+
+    @IBOutlet weak var cellMask: UIView!
+    @IBOutlet weak var bannerImageView: UIImageView!
+    @IBOutlet weak var genderLabel: UILabel!
+    @IBOutlet weak var dayLabel: UILabel!
+    @IBOutlet weak var timeLabel: UILabel!
+    @IBOutlet weak var nameLabel: UILabel!
+    
+    @IBOutlet weak var imageViewAspectRatioConstraint: NSLayoutConstraint!
+    
+    private var currentImageLink: String?
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        
+        self.cellMask.layer.cornerRadius = 20
+        self.cellMask.layer.shadowColor = UIColor.black.cgColor
+        self.cellMask.layer.shadowOpacity = 0.2
+        self.cellMask.layer.shadowOffset = CGSize.zero
+        self.cellMask.layer.shadowRadius = 10
+        self.bannerImageView.layer.cornerRadius = 20
+        self.bannerImageView.layer.maskedCorners = [.layerMaxXMinYCorner, .layerMinXMinYCorner]
+        self.bannerImageView.layer.masksToBounds = true
+    }
+    
+    func configure(with communityGroup: CommunityGroup) {
+        self.genderLabel.text = communityGroup.gender.rawValue.uppercased()
+        self.dayLabel.text = communityGroup.weekDay.rawValue.uppercased()
+        self.timeLabel.text = communityGroup.time
+        self.nameLabel.text = communityGroup.name
+        self.currentImageLink = communityGroup.imageLink
+        self.setImageVisibility(communityGroup.imageLink != nil && !communityGroup.imageLink!.isEmpty)
+        // Try to download the image, but only display it if this cell has not been reused
+        if let imageLink = communityGroup.imageLink {
+            ImageManager.instance.fetch(imageLink) { [weak self] image in
+                if let currentImageLink = self?.currentImageLink, currentImageLink == imageLink {
+                    DispatchQueue.main.async {
+                        self?.bannerImageView.image = image
+                    }
+                }
+            }
+        }
+    }
+    
+    private func setImageVisibility(_ isVisible: Bool) {
+        if isVisible {
```

Comment:
Why is it better to activate the aspect ratio in code rather than in the storyboard?

Reply:
After going back through the code, I realize that this method is called only if an image is provided for the community group. If and only if it is, then the image will be set up with the right aspect ratio. Am I correct?

Reply:
This is definitely a kinda weird piece of code. Essentially, I want to hide the UIImageView if an image doesn't exist in the database, and make sure the UIImageView is visible if the image does exist. However, there are a few problems (keep in mind cell reusability):
- I can't just make the UIImageView invisible because it would still take up the same amount of space on screen, causing a large empty area
- I can't set a height constraint which I programmatically change because I want the image's height to scale depending on the screen size
- I thought I could just deactivate and reactivate a constraint to make it disappear and reappear, but when deactivating a constraint, the constraint gets removed from the UIImageView (and since I have a weak reference to it, the constraint instance is deallocated from memory, setting the variable to nil). If I use a strong reference to the constraint, it would make the code much nicer, but it could also cause a strong retain cycle (memory leak).

What I ended up doing is deactivating the constraint if it exists and I don't want to display an image, and recreating it if it doesn't exist and I do want to display an image. The problem is that now the created constraint will have to be the same as the constraint in Storyboard, which could lead to inconsistencies in the future if only the Storyboard constraint is changed.

Another option would be to use a default image, but if most CommunityGroups don't have images, then I don't think a default image would be ideal

Reply:
Great sounds good! That all makes sense! I like the current solution to the problem and also agree that having a default placeholder image would not be super ideal.